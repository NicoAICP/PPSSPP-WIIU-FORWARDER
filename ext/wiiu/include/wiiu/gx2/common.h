#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
