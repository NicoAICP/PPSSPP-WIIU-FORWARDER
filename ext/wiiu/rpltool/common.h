#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
