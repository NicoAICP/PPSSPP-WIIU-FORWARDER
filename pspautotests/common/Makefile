TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
