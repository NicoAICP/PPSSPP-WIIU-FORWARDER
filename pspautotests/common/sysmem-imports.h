extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();