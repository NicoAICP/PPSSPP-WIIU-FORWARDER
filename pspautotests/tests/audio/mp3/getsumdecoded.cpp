#include <common.h>
#include <pspiofilemgr.h>
#include <pspmp3.h>
#include <psputility.h>
#include <psputils.h>

extern "C" int sceMp3LowLevelInit(int handle, int unk);
extern "C" int sceMp3LowLevelDecode(int handle, const void *src, int *srcConsumed, short *samples, int *sampleBytesWritten);

static u8 mp3Buf[8192] __attribute__((aligned(64)));
static short pcmBuf[4608] __attribute__((aligned(64)));

#define CHECK_ERROR(a) checkError(#a, a);

static int checkError(const char *call, int result) {
	if (result < 0) {
		schedf("ERROR: Unexpected error %08x from %s\n", result, call);
	}
}

static void testGetSumDecoded(const char *title, int handle) {
	int result = sceMp3GetSumDecodedSample(handle);
	if (result >= 0) {
		checkpoint("%s: %d", title, result);
	} else {
		checkpoint("%s: %08x", title, result);
	}
}

static void checkSumLoopChange(const char *title, int handle, int &sum, int &loop) {
	int newSum = sceMp3GetSumDecodedSample(handle);
	int newLoop = sceMp3GetLoopNum(handle);

	if (newSum < sum && newLoop != loop) {
		checkpoint("%s: loop at %d, sum %d => %d", title, newLoop, sum, newSum);
	} else if (newSum < sum) {
		checkpoint("%s: sum at %d => %d", title, sum, newSum);
	} else if (newLoop != loop) {
		checkpoint("%s: loop at %d", title, newLoop);
	}

	sum = newSum;
	loop = newLoop;
}

extern "C" int main(int argc, char *argv[]) {
	sceUtilityLoadModule(PSP_MODULE_AV_AVCODEC);
	sceUtilityLoadModule(PSP_MODULE_AV_MP3);

	sceMp3InitResource();

	int fd = sceIoOpen("sample.mp3", PSP_O_RDONLY, 0777);

	u8 *dst;
	SceInt32 towrite, srcpos;
	short *out = NULL;
	int handle;

	SceMp3InitArg mp3Init;
	mp3Init.mp3StreamStart = 0;
	mp3Init.mp3StreamEnd = sceIoLseek32(fd, 0, SEEK_END);
	mp3Init.unk1 = 0;
	mp3Init.unk2 = 0;
	mp3Init.mp3Buf = mp3Buf;
	mp3Init.mp3BufSize = sizeof(mp3Buf);
	mp3Init.pcmBuf = pcmBuf;
	mp3Init.pcmBufSize = sizeof(pcmBuf);

	checkpointNext("Handles");
	testGetSumDecoded("  Unreserved", 0);
	handle = sceMp3ReserveMp3Handle(NULL);
	testGetSumDecoded("  NULL info arg", handle);
	sceMp3ReleaseMp3Handle(handle);
	handle = sceMp3ReserveMp3Handle(&mp3Init);
	testGetSumDecoded("  Basic info arg", handle);
	sceMp3ReleaseMp3Handle(handle);
	testGetSumDecoded("  Negative", -1);
	testGetSumDecoded("  2", 2);

	checkpointNext("After decoding");
	handle = sceMp3ReserveMp3Handle(NULL);
	sceMp3LowLevelInit(handle, 0);
	testGetSumDecoded("  Low level init", handle);
	sceIoLseek32(fd, 0, SEEK_SET);
	sceIoRead(fd, mp3Buf, 4096);
	int consumed, written;
	sceMp3LowLevelDecode(handle, mp3Buf, &consumed, pcmBuf, &written);
	testGetSumDecoded("  Low level decode", handle);
	sceMp3ReleaseMp3Handle(handle);
	handle = sceMp3ReserveMp3Handle(&mp3Init);
	if (sceMp3GetInfoToAddStreamData(handle, &dst, &towrite, &srcpos) >= 0) {
		sceIoLseek32(fd, srcpos, SEEK_SET);
		sceMp3NotifyAddStreamData(handle, sceIoRead(fd, dst, towrite));
		sceMp3Init(handle);
		testGetSumDecoded("  After init", handle);
		sceMp3Decode(handle, &out);
		testGetSumDecoded("  After decode 1", handle);
		sceMp3Decode(handle, &out);
		testGetSumDecoded("  After decode 2", handle);
	}
	sceMp3ReleaseMp3Handle(handle);

	checkpointNext("Channels");
	static const char *channels[] = { "stereo", "joint stereo", "dual channel", "mono" };
	for (u32 bits = 0; bits < ARRAY_SIZE(channels); ++bits) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Channel %s", channels[bits]);

		handle = sceMp3ReserveMp3Handle(&mp3Init);
		sceMp3GetInfoToAddStreamData(handle, &dst, &towrite, &srcpos);
		sceIoLseek32(fd, srcpos, SEEK_SET);
		sceMp3NotifyAddStreamData(handle, sceIoRead(fd, dst, towrite));

		// Overwrite the channel bits.
		dst[3] = (dst[3] & ~(3 << 6)) | bits << 6;

		sceMp3Init(handle);
		sceMp3Decode(handle, &out);
		testGetSumDecoded(temp, handle);
		sceMp3ReleaseMp3Handle(handle);
	}

	checkpointNext("Loop");
	handle = sceMp3ReserveMp3Handle(&mp3Init);
	sceMp3SetLoopNum(handle, 2);
	int lastSum = 0, lastLoop = 2;
	for (int step = 0; step < 5000; ++step) {
		CHECK_ERROR(sceMp3GetInfoToAddStreamData(handle, &dst, &towrite, &srcpos));
		checkSumLoopChange("  After get add info", handle, lastSum, lastLoop);
		if (towrite != 0) {
			sceIoLseek32(fd, srcpos, SEEK_SET);
			CHECK_ERROR(sceMp3NotifyAddStreamData(handle, sceIoRead(fd, dst, towrite)));
			checkSumLoopChange("  After add", handle, lastSum, lastLoop);
		}
		if (step == 0) {
			CHECK_ERROR(sceMp3Init(handle));
		}
		int decoded = CHECK_ERROR(sceMp3Decode(handle, &out));
		if (decoded <= 0) {
			checkSumLoopChange("  After decode finish", handle, lastSum, lastLoop);
			break;
		} else {
			checkSumLoopChange("  After decode", handle, lastSum, lastLoop);
		}
		if (step == 4999) {
			checkpoint("ERROR: Should have looped by now");
		}
	}
	testGetSumDecoded("  End of loops", handle);

	checkpointNext("Start pos");
	mp3Init.mp3StreamStart = 1;
	handle = sceMp3ReserveMp3Handle(&mp3Init);
	sceMp3SetLoopNum(handle, 2);
	lastSum = 0;
	lastLoop = 2;
	for (int step = 0; step < 5000; ++step) {
		CHECK_ERROR(sceMp3GetInfoToAddStreamData(handle, &dst, &towrite, &srcpos));
		checkSumLoopChange("  After get add info", handle, lastSum, lastLoop);
		if (towrite != 0) {
			sceIoLseek32(fd, srcpos, SEEK_SET);
			CHECK_ERROR(sceMp3NotifyAddStreamData(handle, sceIoRead(fd, dst, towrite)));
			checkSumLoopChange("  After add", handle, lastSum, lastLoop);
		}
		if (step == 0) {
			CHECK_ERROR(sceMp3Init(handle));
		}
		int decoded = CHECK_ERROR(sceMp3Decode(handle, &out));
		if (decoded <= 0) {
			checkSumLoopChange("  After decode finish", handle, lastSum, lastLoop);
			break;
		} else {
			checkSumLoopChange("  After decode", handle, lastSum, lastLoop);
		}
		if (step == 4999) {
			checkpoint("ERROR: Should have looped by now");
		}
	}
	testGetSumDecoded("  End of offset start", handle);

	handle = sceMp3ReserveMp3Handle(&mp3Init);
	sceMp3TermResource();
	checkpointNext("After term");
	testGetSumDecoded("  Prev allocated handle", handle);

	return 0;
}
