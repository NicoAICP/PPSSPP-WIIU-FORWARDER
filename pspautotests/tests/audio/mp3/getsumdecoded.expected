[x] Handles
[x]   Unreserved: 80671102
[x]   NULL info arg: 80671102
[x]   Basic info arg: 0
[x]   Negative: 80671001
[x]   2: 80671001

[x] After decoding
[r]   Low level init: 80671102
[r]   Low level decode: 80671102
[r]   After init: 0
[r]   After decode 1: 1152
[r]   After decode 2: 2304

[x] Channels
[r]   Channel stereo: 1152
[r]   Channel joint stereo: 1152
[r]   Channel dual channel: 1152
[r]   Channel mono: 1152

[x] Loop
[r]   After decode: loop at 1, sum 251136 => 0
[r]   After decode: loop at 0, sum 251136 => 0
[r]   End of loops: 252288

[x] Start pos
[r]   After decode: loop at 1, sum 249984 => 0
[r]   After decode: loop at 0, sum 249984 => 0
[r]   End of offset start: 251136

[x] After term
[x]   Prev allocated handle: 80671001
